File: WorkHub.Application/Authorization/Permissions.cs
namespace WorkHub.Application.Authorization;

public static class Permissions
{
    public static class Organizations
    {
        public const string View     = "organizations.view";
        public const string Update   = "organizations.update";
        public const string Delete   = "organizations.delete";
        public const string ViewBilling = "organizations.billing.view";
    }

    public static class Members
    {
        public const string View     = "members.view";
        public const string Invite   = "members.invite";
        public const string Remove   = "members.remove";
        public const string ChangeRole = "members.changerole";
    }

    public static class Projects
    {
        public const string View     = "projects.view";
        public const string Create   = "projects.create";
        public const string Update   = "projects.update";
        public const string Delete   = "projects.delete";
        public const string Archive  = "projects.archive";
    }

    public static class Tasks
    {
        public const string View     = "tasks.view";
        public const string Create   = "tasks.create";
        public const string Update   = "tasks.update";
        public const string Delete   = "tasks.delete";
        public const string Assign   = "tasks.assign";
        public const string UpdateOwn = "tasks.update.own";
    }

    public static class Comments
    {
        public const string View     = "comments.view";
        public const string Create   = "comments.create";
        public const string Delete   = "comments.delete";
        public const string DeleteOwn = "comments.delete.own";
    }

    public static class Reports
    {
        public const string View     = "reports.view";
        public const string Export   = "reports.export";
    }
}

File: workHub-client/src/lib/permissions.ts
export const Permissions = {
  Org: {
    View:        'organizations.view',
    Update:      'organizations.update',
    Delete:      'organizations.delete',
    BillingView: 'organizations.billing.view',
  },

  Members: {
    View:       'members.view',
    Invite:     'members.invite',
    Remove:     'members.remove',
    ChangeRole: 'members.changerole',
  },

  Projects: {
    View:    'projects.view',
    Create:  'projects.create',
    Update:  'projects.update',
    Delete:  'projects.delete',
    Archive: 'projects.archive',
  },

  Tasks: {
    View:      'tasks.view',
    Create:    'tasks.create',
    Update:    'tasks.update',
    Delete:    'tasks.delete',
    Assign:    'tasks.assign',
    UpdateOwn: 'tasks.update.own',
  },

  Comments: {
    View:      'comments.view',
    Create:    'comments.create',
    Delete:    'comments.delete',
    DeleteOwn: 'comments.delete.own',
  },

  Reports: {
    View:   'reports.view',
    Export: 'reports.export',
  },
} as const

export type Permission =
  | typeof Permissions.Org[keyof typeof Permissions.Org]
  | typeof Permissions.Members[keyof typeof Permissions.Members]
  | typeof Permissions.Projects[keyof typeof Permissions.Projects]
  | typeof Permissions.Tasks[keyof typeof Permissions.Tasks]
  | typeof Permissions.Comments[keyof typeof Permissions.Comments]
  | typeof Permissions.Reports[keyof typeof Permissions.Reports]

export const RolePermissions: Record<string, string[]> = {
  Owner: [
    'organizations.view', 'organizations.update',
    'organizations.delete', 'organizations.billing.view',
    'members.view', 'members.invite', 'members.remove', 'members.changerole',
    'projects.view', 'projects.create', 'projects.update',
    'projects.delete', 'projects.archive',
    'tasks.view', 'tasks.create', 'tasks.update', 'tasks.delete',
    'tasks.assign', 'tasks.update.own',
    'comments.view', 'comments.create', 'comments.delete', 'comments.delete.own',
    'reports.view', 'reports.export',
  ],
  Admin: [
    'organizations.view',
    'members.view', 'members.invite', 'members.remove',
    'projects.view', 'projects.create', 'projects.update', 'projects.delete',
    'tasks.view', 'tasks.create', 'tasks.update', 'tasks.delete',
    'tasks.assign', 'tasks.update.own',
    'comments.view', 'comments.create', 'comments.delete', 'comments.delete.own',
    'reports.view',
  ],
  Member: [
    'organizations.view',
    'members.view',
    'projects.view',
    'tasks.view', 'tasks.update.own',
    'comments.view', 'comments.create', 'comments.delete.own',
  ],
}
